BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch

```diff
@@ -0,0 +1,258 @@
++                ClassLoader classLoader = handlerMethod.getDeclaringClass().getClassLoader();
++                String className = generateNewClassName();
++                byte[] generatedExc = generateEventExecutor(className, handlerMethod);
++                Class<? extends EventExecutor> eventExcClass = defineClass(classLoader, className, generatedExc).asSubclass(EventExecutor.class);
++
++                // Instantiate generated executor
++                return eventExcClass.getConstructor().newInstance();
++            } catch (Exception e) {
++                throw new RuntimeException("Failed to generate an event executor", e);
++            }
++        });
++    }
++
++    /**
++     * @return New ASM event executor class name
++     */
++    static String generateNewClassName() {
++        return ASMEventExecutor.class.getPackage().getName() + ".generated.GeneratedASMEventExecutor" + CLASS_NAME_COUNTER.getAndIncrement();
++    }
++
++    static byte[] generateEventExecutor(String className, Method handlerMethod) {
++        String rawName = className.replace('.', '/');
++
++        Class<?> listenerClass = handlerMethod.getDeclaringClass();
++        Class<?> eventClass = handlerMethod.getParameterTypes()[0];
++
++        String objectName = Type.getInternalName(Object.class);
++        String objectDesc = "L" + objectName + ";";
++        String eeName = Type.getInternalName(EventExecutor.class);
++        String eeDesc = "(" + objectDesc + objectDesc + ")V";
++
++        ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_FRAMES | ClassWriter.COMPUTE_MAXS);
++        cw.visit(Opcodes.V1_8, Opcodes.ACC_PUBLIC, rawName, null, objectName, new String[]{ eeName });
++
++        // Generate no-args public constructor
++        {
++            MethodVisitor mv = cw.visitMethod(Opcodes.ACC_PUBLIC, "<init>", "()V", null, null);
++            mv.visitVarInsn(Opcodes.ALOAD, 0);
++            mv.visitMethodInsn(Opcodes.INVOKESPECIAL, objectName, "<init>", "()V", false);
++            mv.visitInsn(Opcodes.RETURN);
++            mv.visitMaxs(0, 0);
++            mv.visitEnd();
++        }
++
++        // Generate event executor method
++        {
++            MethodVisitor mv = cw.visitMethod(Opcodes.ACC_PUBLIC, "execute", eeDesc, null, null);
++            mv.visitVarInsn(Opcodes.ALOAD, 1);
++            mv.visitTypeInsn(Opcodes.CHECKCAST, Type.getInternalName(listenerClass));
++            mv.visitVarInsn(Opcodes.ALOAD, 2);
++            mv.visitTypeInsn(Opcodes.CHECKCAST, Type.getInternalName(eventClass));
++            int inv = listenerClass.isInterface() ? Opcodes.INVOKEINTERFACE : Opcodes.INVOKEVIRTUAL;
++            mv.visitMethodInsn(inv, Type.getInternalName(listenerClass), handlerMethod.getName(), Type.getMethodDescriptor(handlerMethod), listenerClass.isInterface());
++            if(handlerMethod.getReturnType() != void.class) {
++                mv.visitInsn(Opcodes.POP);
```

Comment:
needs pop2 for long/double

Reply:
Does EventBus even consider methods with non-void returns as event handlers? Then could simplify that logic and remove POP entirely.

Reply:
EventBus does not check return value of @EventHandler methods: https://github.com/SpigotMC/BungeeCord/blob/b85df4f2a19222bf96b9143de38438a95911eec4/event/src/main/java/net/md_5/bungee/event/EventBus.java#L61

Reply:
Where is the facepalm emoji in GitHub when I need it.